src/c-api.adoc

```diff
@@ -746,13 +746,18 @@ statements, including both RISC-V specific and common operand constraints.
 |K                  |5-bit unsigned immediate integer operand     |
 |J                  |Zero integer immediate operand               |
 |s                  |symbol or label reference with a constant offset |
+|cr                 |RVC general purpose register (`x8`-`x15`)  |
+|cf                 |RVC floating point register (`f8`-`f15`)   |
+|Pr                 |Even-odd general purpose register pair     |
```

Comment:
Do any other targets support pair? I haven't been able to find any in LLVM and I don't know how to implement it in LLVM.

Reply:
AArch64 supports tuples of 8 64-bit registers, as implemented here https://github.com/llvm/llvm-project/commit/7d940432c46be83b8fcb5dbefee439585fa820cd

I noticed that the NXP LLVM fork with Zilsd support seems to allocate both `v2i<xlen>` and `i<2*xlen>` to the paired register class, but maybe we only need to do the former? I'm not sure, as it would be nice to be able to pass `uint<2*xlen>_t` to a GPR pair constraint.

I understand that pair constraint is the most complex part of this proposal. I have most of a patch for the other constraints, which I will try to finish in the next few days and post to LLVM.

Reply:
I looked this up later, and on LLVM, SystemZ also supports GPR pairs using `r` - https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/SystemZ/SystemZISelLowering.h#L444-L446

Reply:
I went looking at this and honestly, using a specific constraint rather than something implicit in a modifier seems much more obvious. But thanks for pointing out they had it, the LLVM implementation (they do a bunch of fixing things up later in isel rather than in lowering) was so strange that I hadn't come across what was going on. The RISC-V implementation has ended up following what SystemZ did, which has some overlaps with how Arm works but not many.

Reply:
GCC can't use either `P` or `p` as prefix for constraint due to implementation issue, `P` can only used for integer immediate value only, and `p` can only used for holding address value.

I tried to remove that limitation but I found there is several internal checking are rely on the rule...so I propose "RPr" instead of "Pr"...

NOTE: I have primary implementation which cover all of this PR except using "RPr" rather than "Pr".

[1] https://gcc.gnu.org/onlinedocs/gccint/Simple-Constraints.html

Reply:
BTW, GCC must use constraint to implement register pair, modifier is impossible to implement on GCC side.

Reply:
I think `RPr` is a bit of a mouthful, when most constraints are one letter (and I haven't entirely got my head around the fact that one can provide multiple alternative constraints for the compiler to choose). How about your original suggestion of `R` - this is still "extensible" in that we should be able to use `F` for FPR pairs if we ever need to, and things like `cR` if we ever feel we need to.

I can update the wording here, and updating my LLVM PR won't take too long.

Reply:
'F' is not available on GCC side, it reserved for immediate floating operand...

Reply:
Also one more constraint on GCC side, a multi-letter constraint must have same length, e.g. define both `Rp` and `Rpr` are invalid, since it come with length 2 and 3.

Reply:
I'm starting to wonder how extensible we really need pairs to be. So far they've only turned up in the ISA for GPRs. I think for vector tuples, i think right now it's possible to just use `vr` or `vd` with the right type of argument?

So maybe all we need is GPR pairs, and we don't need to think about extensibility?

Reply:
I've pushed an update with `R`, just saying that `c*` is extensible and pairs are not. We can still change the letter though.

Reply:
> i think right now it's possible to just use `vr` or `vd` with the right type of argument?

Yes, just `vr` or `vd` for vector tuple is fine.

Reply:
> So maybe all we need is GPR pairs, and we don't need to think about extensibility?

I am OK with `R` for GPR pair only for now :)